### PR TITLE
enabled handling of just eight LHEScaleWeights

### PR DESCRIPTION
**Pull request by pkausw**
The current version of the cms LHEScaleWeight producer only allows for exactly 9 weights per event. However, there are also (some) samples that only provide 8 weights (nominal weight is dropped). This leads to a break down of the murf_scale weight producer.

This small PR extends the functionality to also handle samples with only 8 LHE weights. The producer now checks for either 9 or 8 weights. If there are only 8 LHE weights, the indices for the LHE weights that are "normally" behind the nominal weight (i.e. anything after index `4`) are shifted by one to correct the mapping to the corresponding variations.

Side note: The current documentation of NanoAOD v11 does not provide anything about the order of the LHEScaleWeights. Indeed, in the [latest](https://cms-nanoaod-integration.web.cern.ch/autoDoc/NanoAODv11/2022postEE/doc_WZ_TuneCP5_13p6TeV_pythia8_Run3Summer22EENanoAODv11-126X_mcRun3_2022_realistic_postEE_v1-v1.html) version of the documentation, the LHEScaleWeight branch is completely missing from the list of available branches. Might be worthwhile to keep an eye on this.

### CODE REVIEW

**Review by mafrahm — COMMENTED** [review]
Looks good to me! Just one inline comment

File: columnflow/production/cms/scale.py [hunk 1]
```diff
@@ -43,14 +43,39 @@ def murmuf_weights(self: Producer, events: ak.Array, **kwargs) -> ak.Array:
...
+    murf_nominal = None
+    if ak.all(n_weights == 9):
+        # if we have 9 weights, the indices above are correct, just need
+        # to load the nominal weights
+        murf_nominal = events.LHEScaleWeight[:, 4]
+    elif ak.all(n_weights == 8):
+        # if we just have 8 weights, there is no nominal LHEScale weight
+        # instead, initialize the nominal weights as ones.
+        # Additionally, we need to shift the last couple of weight indices
+        # down by 1
```
[the remark this answers is not in the document]
Comment (pkausw) [reply columnflow/production/cms/scale.py:65]: Great suggestion, I commited it directly :+1:

**Comment by pkausw** [comment]
@mafrahm can you give the final approval for this PR?

File: columnflow/production/cms/scale.py [hunk 2]
```diff
@@ -43,14 +43,44 @@ def murmuf_weights(self: Producer, events: ak.Array, **kwargs) -> ak.Array:
...
+            "8 instead of the expected 9. It is assumed, that the missing entry is the " +
+            "nominal one and all other entries are in correct order",
+        )
+        murf_nominal = ak.ones_like(events.event)
+        mur_nom_muf_up -= 1
+        mur_up_muf_down -= 1
+        mur_up_muf_nom -= 1
+        mur_up_muf_up -= 1
+    else:
+
```
Comment (riga) [inline columnflow/production/cms/scale.py:77]: Stray line :)

**Review by riga — CHANGES_REQUESTED** [review]
Hey there! Looks good! 🎉

Just a minor thing regarding "externalizing" the weight lookup indices (with Run 3 nano scheme in mind -.-).

**Comment by pkausw** [comment]
Thanks for your feedback @riga ! Indeed, Mathis and I were discussing whether we should already externalize and stream-line the definition of the weight indices, but in the end decided to postpone this. Your comments are of course valid, especially regarding the up-comping Run 3!

I have started to externalize the index definition of the weights. It's working for the "normal" scale weights, but I haven't tackled the envelope producer yet. It would be nice to generalize the index definition since they are the same for both producers. But I'll have to think about this for a bit. If you already have ideas how best to do this let me know!

**Comment by riga** [comment]
> I have started to externalize the index definition of the weights. It's working for the "normal" scale weights, but I haven't tackled the envelope producer yet. It would be nice to generalize the index definition since they are the same for both producers. But I'll have to think about this for a bit. If you already have ideas how best to do this let me know!

I added a commit with a draft plus minor optimizations (i.e. using a plain `1` instead of `ak.ones_like` if possible and I moved the `ak.any(nominal != 1)` check to the block where this is possible) Ok like this?

Maybe @mafrahm can do the "official" signing then ;)

**Review by mafrahm — APPROVED** [review]
Looks good! Tests in my analysis repository were also successful!

**Pull request merged**